left-sidebar: Hide DMs with deactivated users by default

santhoshh-kumar opened the pull request:
<!-- Describe your pull request here.-->

Fixes: #35102 -- Hide DMs with deactivated users in the left sidebar

This pull request addresses the issue of conversations with deactivated users cluttering the main direct message list.

Previously, any conversation, even those including a deactivated user, would be displayed in the left sidebar. This change introduces a filter to hide these conversations from the default view, providing a cleaner user experience.

The fix is implemented as follows:

- The `pm_list_data.ts` component was updated to fetch all conversations (including those with deactivated users) and then conditionally display them. Deactivated user conversations are hidden from the main list but are still available under the "More conversations" section.
- A new `has_deactivated_user` flag is passed to the `pm_list.ts` component to ensure the "More conversations" link is displayed correctly even if the visible list is short but hidden conversations exist.


<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**


**With deactivated users:**

| Before | After |
| --- | --- |
| <img width="480" height="270" alt="Screenshot from 2025-07-31 18-19-14" src="https://github.com/user-attachments/assets/5a5a560c-9461-4fe4-811a-037678b0cbc6" /> | <img width="480" height="270" alt="Screenshot from 2025-07-31 18-19-43" src="https://github.com/user-attachments/assets/ef01c266-3d77-45e3-8190-008553ae6638" /> |
| -  | <img width="480" height="275" alt="Screenshot from 2025-07-31 18-20-01" src="https://github.com/user-attachments/assets/45602ebc-18d2-4dc2-bd4a-28d648fbedf4" /> |


**Without deactivated users and conversations length exceeding max amount to be shown:**

| Before | After |
| --- | --- |
| <img width="464" height="313" alt="Screenshot from 2025-07-31 18-27-48" src="https://github.com/user-attachments/assets/f547e581-7624-44e3-8ccf-d92590c4d1a2" /> | <img width="464" height="313" alt="Screenshot from 2025-07-31 18-27-48" src="https://github.com/user-attachments/assets/f547e581-7624-44e3-8ccf-d92590c4d1a2" /> |
| <img width="530" height="421" alt="Screenshot from 2025-07-31 18-28-02" src="https://github.com/user-attachments/assets/788258f3-9359-43c2-bdaf-9a236a502f20" /> |  <img width="530" height="421" alt="Screenshot from 2025-07-31 18-28-02" src="https://github.com/user-attachments/assets/788258f3-9359-43c2-bdaf-9a236a502f20" />  |


**Without deactivated users and conversations length less than max amount to be shown:**

| Before | After |
| --- | --- |
| <img width="530" height="351" alt="Screenshot from 2025-07-31 19-05-55" src="https://github.com/user-attachments/assets/1ef71ed9-ea7e-473c-a2d6-a1699e20ee1a" /> | <img width="530" height="351" alt="Screenshot from 2025-07-31 19-05-55" src="https://github.com/user-attachments/assets/1ef71ed9-ea7e-473c-a2d6-a1699e20ee1a" /> |


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review.

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
@Aditya8840 Are you up for doing the first round of review on this one?

santhoshkumarCodes commented on web/src/pm_list_data.ts:
```diff
@@ -88,9 +93,8 @@ export function get_conversations(search_string = ""): DisplayObject[] {
-            Number.parseInt(user_ids_string, 10) || 0,
-        );
+        const user_ids = people.user_ids_string_to_ids_array(user_ids_string);
+        const is_deactivated = user_ids.some((id) => !people.is_active_user_for_popover(id));
```
This change has been made to include both one-on-one and group DMs.

timabbott replied:
OK. `is_active_user_for_popover` should probably be renamed to `is_active_user_or_system_bot` for clarity after this PR :)

We should also call this variable `includes_deactivated_user`, since that's what it is checking.

pratik-pc commented:
You need to add tests for said changes in the same commit instead of separate test commit. Individual commits should pass test coverage.

pratik-pc commented on web/src/pm_list.ts:
```diff
@@ -56,8 +56,9 @@ export function _build_direct_messages_list(opts: {
     );
     const pm_list_info = pm_list_data.get_list_info(zoomed, opts.search_term);
     const more_conversations_unread_count = pm_list_info.more_conversations_unread_count;
+    const has_deactivated_user = pm_list_info.has_deactivated_user;
 
-    if (!opts.all_conversations_shown) {
+    if (!zoomed && (!opts.all_conversations_shown || has_deactivated_user)) {
         pm_list_nodes.push(
             pm_list_dom.more_private_conversations_li(more_conversations_unread_count),
```
I don't think we need `has_deactivated_user`, `all_conversations_shown` should be able to do the same thing. What do you think?

amanagr commented:
Changes look good to me! Tested manually too.

timabbott commented on web/src/pm_list_data.ts:
```diff
@@ -178,12 +185,17 @@ export function get_list_info(
         }
+
+        if (!conversation.is_deactivated) {
+            visible_conversations_count += 1;
```
Why does this happen even when `should_show_conversation` is false? The conversation isn't visible, after all.

timabbott commented on web/e2e-tests/user-deactivation.test.ts:
```diff
@@ -59,6 +59,9 @@ async function test_deactivate_user(page: Page): Promise<void> {
     await page.click(".micromodal .dialog_submit_button");
+    await page.waitForSelector(cordelia_user_row + ".deactivated_user", {
+        visible: true,
+    });
```
What's the story with this new wait in the tests? Probably deserves a comment on what we're waiting for, logically.

santhoshh-kumar replied:
Apologies! I have added it as a seperate commit before - explaining why we need it. But they told me to make everything as a single commit.

The main reason is: ` We need to verify the UI has updated to make sure the user_deactivation_test relying on it to not flake.` 

I will add a comment in code.

timabbott commented on web/tests/pm_list_data.test.cjs:
```diff
@@ -424,3 +424,35 @@ test("get_list_info_no_unread_messages", ({override}) => {
+    assert.deepEqual(
+        list_info.conversations_to_be_shown.map((conversation) => conversation.recipients),
+        ["Alice, Cardelio", "Cardelio, Zoe", "Cardelio", "Zoe", "Me Myself", "Alice"],
+    );
```
It'd be good if this test had more conversations available than the number that actually get shown, so that one can verify the behavior around https://github.com/zulip/zulip/pull/35563/files#r2500834116.

timabbott commented:
Re-assigned to @amanagr for further maintainer review.

amanagr commented:
@laurynmm can you please review this since you are working in this area.

laurynmm commented on web/src/pm_list_data.ts:
```diff
@@ -84,8 +84,9 @@ export function get_conversations(search_string = ""): DisplayObject[] {
         const is_active = user_ids_string === active_user_ids_string;
-        const is_deactivated = !people.is_active_user_for_popover(
-            Number.parseInt(user_ids_string, 10) || 0,
+        const user_ids = people.user_ids_string_to_ids_array(user_ids_string);
```
This is now called twice (see where `users` is defined above), so we should probably just get the array once and then use that in both places.

laurynmm commented on web/src/pm_list_data.ts:
```diff
@@ -157,10 +163,16 @@ export function get_list_info(
         // conversations, regardless of whether they have unread
         // messages.
-        if (idx < max_conversations_to_show) {
+        if (visible_conversations_count < max_conversations_to_show) {
```
This can be used in the if statement below as well, correct?

santhoshh-kumar replied:
After hearing your feedback. I don't think we need the addition of `visible_conversations_count` and it's counting logic. We can directly use the `conversations_to_be_shown.length` by defining a function level variable for it.

What do you think?

laurynmm commented on web/tests/pm_list.test.cjs:
```diff
@@ -31,3 +30,106 @@ run_test("update_dom_with_unread_counts", () => {
     assert.equal($total_count.hasClass("hide"), true);
 });
+
+run_test("build_direct_messages_list", ({override}) => {
```
I don't think we need this test addition. It's basically confirming that the boolean value for `all_conversations_shown` determines whether the `more_private_conversations_li` is rendered.

laurynmm reviewed (commented):
@santhoshh-kumar - Thanks for working on this issue!

I had a few minor comments about the code changes, but overall they look good. I think that the changes to the tests can be simplified a bit, which I've noted in a few review comments.

Finally, I think you need to updated your commit message as the changes no longer involve updating `pm_list.ts`. Let me know if you have any questions here or on CZO.

santhoshh-kumar commented:
Changes:
- Removed all the unnecessary test additions.
- Removed the `visible_conversation_count` and added function level variable named `conversation_to_be_shown_count` due to the reason stated [here](https://github.com/zulip/zulip/pull/35563#discussion_r2536371590).
- Refactored to get the `user_ids` array once and use it in both the required places.

santhoshh-kumar commented:
Thanks for those changes!
